karapace/schema_reader.py

```diff
@@ -464,27 +476,93 @@ def _handle_msg_schema(self, key: dict, value: Optional[dict]) -> None:
+            try:
+                if schema_references:
+                    resolved_references = [
+                        Reference(reference["name"], reference["subject"], reference["version"])
+                        for reference in schema_references
+                    ]
+                    resolved_dependencies = self.resolve_references(resolved_references)
+                parsed_schema = parse_protobuf_schema_definition(
+                    schema_str,
+                    resolved_references,
+                    resolved_dependencies,
+                    validate_references=False,
+                )
+                schema_str = str(parsed_schema)
+            except InvalidSchema:
+                LOG.exception("Schema is not valid ProtoBuf definition")
+                return
+
         try:
             typed_schema = TypedSchema(
                 schema_type=schema_type_parsed,
                 schema_str=schema_str,
+                references=resolved_references,
+                dependencies=resolved_dependencies,
+                schema=parsed_schema,
             )
         except (InvalidSchema, JSONDecodeError):
             return
+
         self.database.insert_schema_version(
             subject=schema_subject,
             schema_id=schema_id,
             version=schema_version,
             deleted=schema_deleted,
             schema=typed_schema,
+            references=resolved_references,
         )
 
+        if resolved_references:
+            for ref in resolved_references:
+                self.database.insert_referenced_by(subject=ref.subject, version=ref.version, schema_id=schema_id)
+
     def handle_msg(self, key: dict, value: Optional[dict]) -> None:
         if key["keytype"] == "CONFIG":
             self._handle_msg_config(key, value)
         elif key["keytype"] == "SCHEMA":
+            LOG.error("HANDLING SCHEMA MESSAGE")
             self._handle_msg_schema(key, value)
         elif key["keytype"] == "DELETE_SUBJECT":
             self._handle_msg_delete_subject(key, value)
         elif key["keytype"] == "NOOP":  # for spec completeness
             pass
+
+    def remove_referenced_by(self, schema_id: SchemaId, references: List[Reference]):
+        self.database.remove_referenced_by(schema_id, references)
+
+    def get_referenced_by(self, subject: Subject, version: ResolvedVersion) -> Optional[Referents]:
+        return self.database.get_referenced_by(subject, version)
+
+    def _resolve_reference(self, reference: Reference) -> Dependency:
+        subject_data = self.database.find_subject_schemas(subject=reference.subject, include_deleted=False)
+        if not subject_data:
+            raise InvalidReferences(f"Subject not found {reference.subject}.")
+        schema_version: SchemaVersion = subject_data.get(reference.version, {})
+        if schema_version is None:
+            raise InvalidReferences(f"Subject {reference.subject} has no such schema version")
```

Comment:
This won't be thrown if the `reference.version` is not present in the subject_data -- you mean `subject_data.get(reference.version, None)`?

Reply:
yes